### PR TITLE
i#5442: Fix decoding of the compact (90) form of xchg %ax,%r8w

### PR DESCRIPTION
**Pull request by khuey**
xchg %ax,%r8w is separated from xchg %ax,%ax (aka OP_nop) only by REX.B.
The decoding table was not reaching the 90 OP_xchg when coming through
prefix_extensions.

Encoding of this short variant (and also e.g. xchg %eax, %r8d; xchg %eax, %r9d; etc)
is still not handled, but those will fall back to the longer 87 xchg opcode.

Issue: #5442

### CODE REVIEW

File: core/ir/x86/decode_table.c [hunk 1]
```diff
@@ -4604,10 +4604,9 @@ const instr_info_t prefix_extensions[][12] = {
...
   }, /* prefix extension 103 */
   {
-    {REX_B_EXT,  0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, 0},
+    {REX_B_EXT,  0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, END_LIST},
     {OP_pause,0xf3900000, "pause", xx, xx, xx, xx, xx, no, x, END_LIST},
-    /* we chain these even though encoding won't find them */
-    {OP_nop, 0x66900000, "nop", xx, xx, xx, xx, xx, no, x, tpe[103][3]},
+    {REX_B_EXT, 0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, END_LIST},
     /* windbg displays as "repne nop" */
     {OP_nop, 0xf2900000, "nop", xx, xx, xx, xx, xx, no, x, END_LIST},
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:4611]: This is wrong too:
```
$ disasm f2 41 90
Disassembling 0xf2 0x41 0x90
llvm-mc:   0xf2 0x41 0x90  repne xchgl %r8d, %eax
capstone:  41f2 xchg eax, r8d
bfd:       f2 41 90 repnz xchg %eax,%r8d
DynamoRIO: f2 41 90 nop
```

File: core/ir/x86/decode_table.c [hunk 2]
```diff
@@ -4604,10 +4604,9 @@ const instr_info_t prefix_extensions[][12] = {
     {INVALID, 0xf20f7337, "(bad)", xx, xx, xx, xx, xx, no, x, NA},
   }, /* prefix extension 103 */
   {
-    {REX_B_EXT,  0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, 0},
+    {REX_B_EXT,  0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, END_LIST},
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:4607]: END_LIST is for a leaf entry; as a next-table entry this should be an index.

File: core/ir/x86/decode_table.c [hunk 3]
```diff
@@ -4604,10 +4604,9 @@ const instr_info_t prefix_extensions[][12] = {
...
     {OP_pause,0xf3900000, "pause", xx, xx, xx, xx, xx, no, x, END_LIST},
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:4608]: We should figure out who is correct here:
```
$ disasm f3 41 90
Disassembling 0xf3 0x41 0x90
llvm-mc:   0xf3 0x41 0x90  xchgl %r8d, %eax
capstone:  41f3 xchg eax, r8d
bfd:       f3 41 90 rex.B pause
DynamoRIO: f3 41 90 pause
```

File: core/ir/x86/decode_table.c [hunk 4]
```diff
@@ -4604,10 +4604,9 @@ const instr_info_t prefix_extensions[][12] = {
...
-    /* we chain these even though encoding won't find them */
-    {OP_nop, 0x66900000, "nop", xx, xx, xx, xx, xx, no, x, tpe[103][3]},
+    {REX_B_EXT, 0x900000, "(rex.b ext 0)", xx, xx, xx, xx, xx, no, x, END_LIST},
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:4609]: I would keep the 0x66 in the opcode field as it makes it easier to understand which row this is.

Comment (derekbruening) [inline core/ir/x86/decode_table.c:4609]: END_LIST is for a leaf entry; as a next-table entry this should be an index.

File: core/ir/x86/decode_table.c [hunk 5]
```diff
@@ -7118,7 +7117,8 @@ const instr_info_t evex_prefix_extensions[][2] = {
  */
 const instr_info_t rex_b_extensions[][2] = {
   { /* rex.b extension 0 */
-    {OP_nop,  0x900000, "nop", xx, xx, xx, xx, xx, no, x, tpe[103][2]},
+    /* we chain these even though encoding won't find them */
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:7120]: style nit: Prefer capitalized and punctuated sentences for whole-line comments.

File: core/ir/x86/decode_table.c [hunk 6]
```diff
@@ -7118,7 +7117,8 @@ const instr_info_t evex_prefix_extensions[][2] = {
...
+    {OP_nop,  0x900000, "nop", xx, xx, xx, xx, xx, no, x, tpe[103][3]},
     /* For decoding we avoid needing new operand types by only getting
      * here if rex.b is set.  For encode, we would need either to take
```
Comment (derekbruening) [inline core/ir/x86/decode_table.c:7123]: So this comment is explaining why the encoder won't find this?  Wondering if we should file a separate issue so we can point at it in the test and PR description even if we don't plan to fix it.

**Review by derekbruening — COMMENTED** [review]
Thank you for the patch!  I think we just need to add the 0xf2 and maybe 0xf3 prefixes too.

File: core/ir/x86/decode_table.c (hunk 4 above)
Comment (khuey) [reply core/ir/x86/decode_table.c:4609]: Adding 0x66 to the opcode field will trigger the logic at https://github.com/DynamoRIO/dynamorio/blob/30da7c1edd19fac69ef61334a90fb5a4df0d17c9/core/ir/x86/decode.c#L916 and result in wrong operand sizes.

File: core/ir/x86/decode_table.c (hunk 3 above)
Comment (khuey) [reply core/ir/x86/decode_table.c:4608]: xed says pause

File: core/ir/x86/decode_table.c (hunk 6 above)
Comment (khuey) [reply core/ir/x86/decode_table.c:7122]: Yes, and done.

File: core/ir/x86/decode_table.c (hunk 3 above)
Comment (derekbruening) [reply core/ir/x86/decode_table.c:4608]: gdb also says pause though I thought it used the bfd code (maybe newer than my objdump):
```
=> 0x40100f <_start+15>:	rex.B pause 
```
A quick asm test shows no xchg happening on my processor at least:
```
(gdb) info reg
rax            0xdeadbeef          3735928559
r8             0x12345678          305419896
(gdb) x/3bx $pc
0x40100f <_start+15>:	0xf3	0x41	0x90
(gdb) x/1i $pc
=> 0x40100f <_start+15>:	rex.B pause 
(gdb) stepi
0x0000000000401012 in _start ()
(gdb) info reg
rax            0xdeadbeef          3735928559
r8             0x12345678          305419896
```

**Review by derekbruening — APPROVED** [review]

**Comment by derekbruening** [comment]
Oh the Jenkins is down (machine being rebuilt) so I will have to override to merge.  Updating now.

**Pull request merged**